sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,539 @@
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(String.class);
+    }
+
+    @Override
+    public void validate() {
+      Preconditions.checkNotNull(uri, "uri");
+      Preconditions.checkNotNull(database, "database");
+      Preconditions.checkNotNull(collection, "collection");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      builder.addIfNotNull(DisplayData.item("uri", uri));
+      builder.addIfNotNull(DisplayData.item("database", database));
+      builder.addIfNotNull(DisplayData.item("collection", collection));
+      builder.addIfNotNull(DisplayData.item("filter", filter));
+      builder.addIfNotNull(DisplayData.item("numSplit", numSplits));
+    }
+
+    @Override
+    public boolean producesSortedKeys(PipelineOptions options) {
+      return false;
+    }
+
+    @Override
+    public BoundedReader createReader(PipelineOptions options) {
+      return new BoundedMongoDbReader(this);
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) {
+      long estimatedSizeBytes = 0L;
+
+      MongoClient mongoClient = new MongoClient();
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+      MongoCollection mongoCollection = mongoDatabase.getCollection(collection);
+
+      // get the Mongo collStats object
+      // it gives the size for the entire collection
+      BasicDBObject stat = new BasicDBObject();
+      stat.append("collStats", collection);
+      Document stats = mongoDatabase.runCommand(stat);
+      estimatedSizeBytes = Long.valueOf(stats.get("size").toString());
+      long avgSize = Long.valueOf(stats.get("avgObjSize").toString());
+      if (filter != null && !filter.isEmpty()) {
+        // as the user is using a filter, we improve the estimated size
+        // by counting number of document returned by the filter
+        // and using the average object size from collStats
+        try {
+          estimatedSizeBytes = mongoCollection.count(Document.parse(filter)) * avgSize;
+        } catch (Exception e) {
+          LOGGER.warn("Can't estimate size", e);
+        }
+      }
+      return estimatedSizeBytes;
+    }
+
+    @Override
+    public List<BoundedSource<String>> splitIntoBundles(long desiredBundleSizeBytes,
+                                                PipelineOptions options) {
+      MongoClient mongoClient = new MongoClient();
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+
+      List<Document> splitKeys = null;
+      if (numSplits > 0) {
+        // the user defines his desired number of splits
+        // calculate the bundle size
+        long estimatedSizeBytes = getEstimatedSizeBytes(options);
+        desiredBundleSizeBytes = estimatedSizeBytes / numSplits;
+      }
+
+      // the desired bundle size is small, using default chunk size of 64MB
+      if (desiredBundleSizeBytes == 0 || ((desiredBundleSizeBytes / 1024 / 1024) < 1)) {
```

Comment:
This can be simplified to "if (desiredBundleSizeBytes < 1024 * 1024)". And then the condition looks weird: why do we adjust desiredBundleSizeBytes under 1MB to 64MB, but, say, 3MB we'll keep as is? Shouldn't this be just desiredBundleSizeBytes = max(desiredBundleSizeBytes, 64MB) ? (I don't know whether or not this is the right thing to do)